prog/dftb+/lib_geoopt/lbfgs.F90

```diff
@@ -0,0 +1,886 @@
+    this%gg(:) = gg
+
+    if (.not. present(diagIn)) then
+      ! Compute guessed hessian matrix if not set by the user
+      if (this%iter == 1) then
+        diag(:) = 1.0_dp
+      else
+        diag(:) = dot_product(this%ss(:, newElem), this%yy(:, newElem))&
+            & / dot_product(this%yy(:, newElem), this%yy(:, newElem))
+      end if
+
+    else if (any(diagIn < 0.0_dp)) then
+      ! If hessian set by the user, do input check
+      call error("Error: Bad Input (diag). (LBFGS Minimizer)")
+    else
+      diag(:) = diagIn
+    end if
+
+    ! Compute d = -H*g using the formula given by Nocedal. Result saved in this%dir
+    qq(:) = gg
+    ii = newElem
+    do
+      alpha(ii) = this%rho(ii) * dot_product(this%ss(:,ii), qq)
+      qq(:) = qq - alpha(ii) * this%yy(:,ii)
+
+      ii = ii - 1
+      if (ii == 0) then
+        ii = this%mem
+      end if
+
+      if (ii == newElem) then
+        exit
+      end if
+
+    end do
+
+    this%dir(:) = diag * qq
+
+    if (newElem + 1 > this%mem) then
+      oldestElem = 1
+    else
+      oldestElem = newElem + 1
+    end if
+    ii = oldestElem
+    do
+      beta = this%rho(ii) * dot_product(this%yy(:,ii), this%dir)
+      this%dir(:) = this%dir + this%ss(:,ii) * (alpha(ii) - beta)
+
+      ii = ii + 1
+      if (ii > this%mem) then
+        ii = 1
+      end if
+
+      if (ii == oldestElem) then
+        exit
+      end if
+
+    end do
+
+    this%iter = this%iter + 1
+
+    ! Change to correct sign
+    this%dir(:) = -this%dir
+
+  end subroutine calcDirection
+
+
+  !> Creates a new line minimizer
+  subroutine TLineMin_init(this, nElem, mIter, tolerance, maxDisp)
+
+    !> Valid line minimizer instance on exit
+    type(TLineSearch), intent(out) :: this
+
+    !> Nr. of elements in the coordinate/gradient vectors
+    integer, intent(in) :: nElem
+
+    !> Nr. of maximal iterations to perform (>3)
+    integer, intent(in) :: mIter
+
+    !> Convergence criteria for the projected derivative
+    real(dp), intent(in) :: tolerance
+
+    !> Maximal movement in one coordinate in one step
+    real(dp), intent(in) :: maxDisp
+
+    @:ASSERT(nElem > 0)
+    @:ASSERT(mIter > 3)
+    @:ASSERT(maxDisp > 0.0_dp)
+
+    this%nElem = nElem
+    this%maxDisp = maxDisp
+    this%mIter = mIter
+
+    this%tol = tolerance
+
+    allocate(this%d0(nElem))
+    allocate(this%dxLo(nElem))
+    allocate(this%dxHi(nElem))
+    allocate(this%xNew(nElem))
+    allocate(this%x0(nElem))
+    allocate(this%dx0(nElem))
+
+  end subroutine TLineMin_init
+
+
+  !> Resets the line minimizer
+  subroutine TLineMin_reset(this, x0, d0, firstStep)
+
+    !> Line minimizer instance
+    type(TLineSearch), intent(inout) :: this
+
+    !> New starting point
+    real(dp), intent(in) :: x0(:)
+
+    !> New direction
+    real(dp), intent(in) :: d0(:)
+
+    real(dp), intent(in) :: firstStep
+
+    @:ASSERT(size(x0) == this%nElem)
+    @:ASSERT(size(d0) == this%nElem)
+    #!@:ASSERT(all(abs(d0) > 0.0_dp))
```

Comment:
I had to disable this ASSERT in order to get the tests working. Shouldn't this be `any()` instead of all?